workflow/controller/operator.go

```diff
@@ -1368,12 +1368,31 @@ func (woc *wfOperationCtx) createPVCs() error {
 	return nil
 }
 
+func (woc *wfOperationCtx) getVolumeGcStrategy() wfv1.VolumeGCStrategy {
+	// If no volumeGC strategy was provided, we default to the equivalent of "OnSuccess"
+	// to match the existing behavior for back-compat
+	if woc.wf.Spec.VolumeGC == nil {
+		return wfv1.VolumeGCOnSuccess
+	}
+
+	return woc.wf.Spec.VolumeGC.Strategy
+}
+
 func (woc *wfOperationCtx) deletePVCs() error {
-	if woc.wf.Status.Phase == wfv1.NodeError || woc.wf.Status.Phase == wfv1.NodeFailed {
-		// Skip deleting PVCs to reuse them for retried failed/error workflows.
-		// PVCs are automatically deleted when corresponded owner workflows get deleted.
+	switch woc.getVolumeGcStrategy() {
+	case wfv1.VolumeGCOnSuccess:
+		if woc.wf.Status.Phase == wfv1.NodeError || woc.wf.Status.Phase == wfv1.NodeFailed {
```

Comment:
why not 'woc.wf.Status.Phase == wfv1.Succeeded`?

Reply:
I had left it that way due to the change made in https://github.com/argoproj/argo/pull/2449, although the original check there was `woc.wf.Status.Phase != wfv1.NodeSucceeded` versus what you had suggested here.

I'm not familiar enough with all of the status fields (yet) to know whether or not that change is safe, and unfortunately that PR didn't include tests for the change, so I erred on the side of safety

Reply:
Actually looking at the code, the workflow is marked as succeeded after PVCs have been deleted, meaning that we can't use that check here, at least according to this block of code:

https://github.com/argoproj/argo/blob/5a0c515c45f8fbcf0811c25774c1c5f97e72286d/workflow/controller/operator.go#L397-L410